aptos-move/aptos-native-interface/src/builder.rs

```diff
@@ -119,32 +107,55 @@ impl SafeNativeBuilder {
                 native_gas_params: &data.native_gas_params,
                 misc_gas_params: &data.misc_gas_params,
 
-                gas_budget,
-                gas_used: 0.into(),
-
-                enable_incremental_gas_charging,
+                legacy_gas_used: 0.into(),
+                legacy_enable_incremental_gas_charging: enable_incremental_gas_charging,
+                legacy_heap_memory_usage: 0,
 
                 gas_hook: hook.as_deref(),
             };
 
             let res: Result<SmallVec<[Value; 1]>, SafeNativeError> =
                 native(&mut context, ty_args, args);
 
+            // If enabled, metering and memory tracking must have been done in the native!
+            let legacy_heap_memory_usage = context.legacy_heap_memory_usage;
+            if context.has_direct_gas_meter_access_in_native_context() {
+                assert_eq!(context.legacy_gas_used, 0.into());
+                assert_eq!(legacy_heap_memory_usage, 0);
+            }
+            context
```

Comment:
@gelash call to use memory happens here: this is a real gas meter, it is just not used for old versions when accumulating memory. In the end, we charge accumulated amount: 0 for new version (no-op), and actual amount for old version.

Reply:
Yes! I had closed this file lol. Took me two tries to even find the tag :D I was looking for this because I remembered even before I asked lol. Thx makes sense.